### PR TITLE
fix: fixes getFanOut; adds test for data analyzer

### LINKED ISSUE #991: How to use the getFanOut?
I'm trying to use the function `getFanout` in `spinal.lib.tools.DataAnalyzer`.  Here is my code:

```scala
import spinal.core._
import spinal.lib._

class MyTry extends Component {
  val io = new Bundle {
    val a = in Bool()
    val b = out Bool()
  }
  io.b:=io.a
}

object MyTry extends App {
    SpinalVerilog(new MyTry)
    val mytest=new MyTry
    val analyzer_a=new DataAnalyzer(mytest.io.a)
    println("fanout signal of a:"+analyzer_a.getFanOut)
    val analyzer_b=new DataAnalyzer(mytest.io.b)
    println("fanin signal of b:"+analyzer_b.getFanIn)
}
```

here is the results

```
fanout signal of a:Set()
fanin signal of b:Set((toplevel/io_a : in Bool))
```

Why the fanout signal of a is null?

### PR DESCRIPTION
**Pull request by wswslzp**
Closes #991 

# Context, Motivation & Description

As the title. 

# Impact on code generation

No impact.

# Checklist

- [x] Unit tests were added
- [x] API changes are or will be documented:
  - using Scaladoc comments: `/** */`?
  - on [RTD](https://github.com/SpinalHDL/SpinalDoc-RTD)?
  - thanks to a [new tracking issue on RTD](https://github.com/SpinalHDL/SpinalDoc-RTD/issues/new?title=Document%20XXX&body=Do%20not%20merge%20until%20SpinalHDL/SpinalHDL%23XXX%20has%20not%20been%20merged)?

### CODE REVIEW

File: tester/src/test/scala/spinal/tester/scalatest/DataAnalyzerTest.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    val a2 = RegNext(a2_1)
+
+    io.b := (a2, a1, a0).asBits
+
+  }
+
+  test("Gen Test Module") {
+    SpinalVerilog(TestModule()).printRtl()
+  }
+
```
Comment (numero-744) [inline tester/src/test/scala/spinal/tester/scalatest/DataAnalyzerTest.scala:26]:
```suggestion
```

I think this test is not needed

File: lib/src/main/scala/spinal/lib/tools/DataAnalyzer.scala [hunk 2]
```diff
@@ -18,18 +18,31 @@ class DataAnalyzer(data: BaseType) {
     */
   def getFanIn: mutable.LinkedHashSet[BaseType] = {
     val ret = mutable.LinkedHashSet.newBuilder[BaseType]
-    data.foreachStatements { st =>
+
+    def traceBaseTypeFromExpression(e: Expression): mutable.LinkedHashSet[BaseType] = {
+      val ret = mutable.LinkedHashSet.newBuilder[BaseType]
+      e.foreachDrivingExpression{
```
Comment (numero-744) [inline lib/src/main/scala/spinal/lib/tools/DataAnalyzer.scala:24]: Can you auto-format this file, please?

**Comment by ZhaokunHu** [comment]
Thank you for the PR and the example ！I understand how to use it now and it works good.

**Comment by wswslzp** [comment]
All done, I think it's ok to be merged.

**Comment by Readon** [comment]
seems that current analyzer is designed for generated rtl, according to the test.
is it applicapable for modules before elaboration? where the signals' name are different.

do you think it might be better if there is an test for that use case?

**Comment by Readon** [comment]
seems that current analyzer is designed for generated rtl, according to the test.
is it applicapable for modules before elaboration? where the signals' name are different.

do you think it might be better if there is an test for that use case?

**Pull request merged**